### PR TITLE
GEOMESA-2335 Refactoring Kafka data store to allow for mock expiration

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Event time/expiration logic moved to FeatureStateFactory implementations
* Consumers not woken up on initial load close, as it can trigger a WakeupException on next poll

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryFeatureStateFactory.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/**
+  * Non-point feature index with event time expiration but message time ordering
+  *
+  * @param index spatial index
+  * @param geom geometry attribute
+  * @param executor executor
+  * @param expiry expiry in millis
+  * @param eventTime event time expression
+  */
+class ExtentsEventTimeExpiryFeatureStateFactory(index: SpatialIndex[SimpleFeature],
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryFeatureStateFactory.scala]: I'm tempted to suggest having these Factories in the same file.  Thoughts?

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/EventTimeFeatureCache.scala [hunk 2]
```diff
@@ -1,100 +0,0 @@
-/***********************************************************************
- * Copyright (c) 2013-2018 Commonwealth Computer Research, Inc.
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/EventTimeFeatureCache.scala]: I moved the event time logic into the FeatureStateFactory classes

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryOrderingFeatureStateFactory.scala [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+/**
+  * Non-point feature index with event time ordering and timed expiration
+  *
+  * @param index spatial index
+  * @param geom geometry attribute
+  * @param executor executor
+  * @param expiry expiry in millis
+  * @param eventTime event time expression
+  */
+class ExtentsEventTimeExpiryOrderingFeatureStateFactory(index: SpatialIndex[SimpleFeature],
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryOrderingFeatureStateFactory.scala]: *blinks*  Ok, so I'm reading quickly... why do some of these have an Executor and others not?

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryFeatureStateFactory.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryFeatureStateFactory.scala]: i originally had them in one file, but it was kind of confusing to read - i don't really mind either way

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryOrderingFeatureStateFactory.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeExpiryOrderingFeatureStateFactory.scala]: the ones that expire features have an executor

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeFeatureStateFactory.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package org.locationtech.geomesa.kafka.index.state
+
+import com.typesafe.scalalogging.LazyLogging
+import org.locationtech.geomesa.kafka.index.state.FeatureStateFactory._
+import org.locationtech.geomesa.utils.index.SpatialIndex
+import org.opengis.feature.simple.SimpleFeature
+import org.opengis.filter.expression.Expression
+
+/**
+  * Non-point feature state with even time ordering
```
Comment (jnh5y) [inline geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeFeatureStateFactory.scala]: Typo: *event* time.

Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/state/ExtentsEventTimeFeatureStateFactory.scala]: thanks, fixed

**Pull request merged**